sampledb/api/server/object_permissions.py

```diff
@@ -3,16 +3,257 @@
+                }, 400
+            permissions = Permissions.from_name(permission)
+            if permissions not in {Permissions.NONE, Permissions.READ}:
+                return {
+                    "message": 'expected "none" or "read"'
+                }, 400
+            object_permissions.set_object_permissions_for_all_users(
+                object_id=object_id,
+                permissions=permissions,
+            )
+
+        # Anonymous
+        if (permission := request_json.get("anonymous_users")) is not None:
+            if permission is not None and not flask.current_app.config['ENABLE_ANONYMOUS_USERS']:
+                return {
+                    "message": "anonymous users are disabled"
+                }, 400
+            if not isinstance(permission, str):
+                return {
+                    "message": "JSON string for permission required"
+                }, 400
+            permissions = Permissions.from_name(permission)
+            if permissions not in {Permissions.NONE, Permissions.READ}:
+                return {
+                    "message": 'expected "none" or "read"'
+                }, 400
+            object_permissions.set_object_permissions_for_anonymous_users(
+                object_id=object_id,
+                permissions=permissions,
+            )
+        return all_object_permissions_dict_to_json(
+            object_permissions.get_all_object_permissions(object_id=object_id)
+        ), 200
+
+
+class CopyObjectsPermissions(Resource):
+    @multi_auth.login_required
+    def post(self) -> ResponseData:
+        user_id = flask.g.user.id
+        user = users.get_user(user_id=user_id)
+        if user.is_readonly:
+            return {
+                "message": "No permission to copy object's permissions"
+            }, 403
+        request_json = flask.request.get_json(force=True)
+        if isinstance(request_json, dict):
+            request_json = [request_json]
+        elif not isinstance(request_json, list):
+            return {
+                "message": "JSON list or dict body required"
+            }, 400
+        try:
+            for item in request_json:
+                if not isinstance(item, dict):
+                    return {
+                        "message": "JSON list or dict body required"
+                    }, 400
+                if source_object_id := item.get("source_object_id", None) is None:
+                    return {
+                        "message": "source_object_id for JSON object required"
+                    }, 400
+                if target_object_id := item.get("target_object_id", None) is None:
+                    return {
+                        "message": "source_object_id for JSON object required"
+                    }, 400
+                # Validating all incoming object ids before changing any permission
+                for object_id in [source_object_id, target_object_id]:
+                    object_id = int(object_id)
```

Comment:
mypy complains about this line: `sampledb/api/server/object_permissions.py:207: error: Incompatible types in assignment (expression has type "int", variable has type "Literal[False]")  [assignment]` because both `source_object_id` and `target_object_id` contains the result of the comparison with `None`, so both are `False` when everything goes well, which will be converted to `0` with `int`, which obviously wasn't your intention. assign `source_object_id` to `item.get("source_object_id")` first and then do the comparison, the same for `target_object_id` and you'll avoid that issue.